README.md
# html_widget

A Flutter package to display html document.

## Getting Started

This package provides a widget to display html document in a Flutter app.

You can use it like this:

```dart 
List<Widget> richText = MyHtmlParser.parseHtmlToListOfTextWidgets(htmlDoc);
```

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/), a library module containing code that can be
shared easily across multiple Flutter or Dart projects.

For help getting started with Flutter, view our
[online documentation](https://flutter.dev/docs), which offers tutorials, samples, guidance on
mobile development, and a full API reference.


CHANGELOG.md
## 0.0.3

* This release only supports the following html tags: h1, h2, p, ul, ol, li, b.


lib/html_parser.dart
library html_parser;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:html/parser.dart';

class MyHtmlParser {
  static List<Widget> parseHtmlToListOfTextWidgets(var text) {
    var data = parse(text.toString()).body;
    List<Widget> richText = [];
    for (int i = 0; i < data!.nodes.length; i++) {
      if (data.nodes[i].toString().contains('<html h1>')) {
        richText.add(_titleWidget(data.nodes[i].text));
      } else if (data.nodes[i].toString().contains('<html h2>')) {
        richText.add(_subtitleWidget(data.nodes[i].text));
      } else if (data.nodes[i].toString().contains('<html p>')) {
        richText.add(_paragraphWidget(data.nodes[i], data.nodes[i].text));
      } else if (data.nodes[i].toString().contains('<html ol>')) {
        richText.add(_numberedListWidget(data.nodes[i].nodes));
      } else if (data.nodes[i].toString().contains('<html ul>')) {
        richText.add(_bulletListWidget(data.nodes[i].nodes));
      } else if (data.nodes[i].toString().contains('<html b>')) {
        richText.add(_boldWidget(data.nodes[i].text));
      }
    }
    return richText;
  }

  static Widget _titleWidget(String? title) {
    return Text(title!,
        style: TextStyle(fontSize: 32, fontWeight: FontWeight.bold));
  }

  static Widget _subtitleWidget(String? subtitle) {
    return Text(subtitle!,
        style: TextStyle(fontSize: 24, fontWeight: FontWeight.bold));
  }

  static Widget _boldWidget(String? text) {
    return Text(text!, style: TextStyle(fontWeight: FontWeight.bold));
  }

  static Widget _paragraphWidget(var nodesList, String? text) {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 8.0),
      child: nodesList.nodes.isEmpty
          ? Text(text!)
          : RichText(
              text: TextSpan(
                style: TextStyle(
                  fontSize: 16.0,
                  color: Colors.black,
                ),
                children: _textSpanList(nodesList.nodes),
              ),
            ),
    );
  }

  static Widget _bulletListWidget(var bulletPoints) {
    return ListView.builder(
        shrinkWrap: true,
        itemCount: bulletPoints.length,
        itemBuilder: (BuildContext context, int index) {
          return bulletPoints[index].nodes.isEmpty
              ? Text('• ' + bulletPoints[index].text)
              : Row(
                  mainAxisAlignment: MainAxisAlignment.start,
                  mainAxisSize: MainAxisSize.max,
                  children: [
                    Text('• '),
                    Expanded(
                      child: RichText(
                        text: TextSpan(
                          style: DefaultTextStyle.of(context).style,
                          children: _textSpanList(bulletPoints[index].nodes),
                        ),
                      ),
                    ),
                  ],
                );
        });
  }

  static Widget _numberedListWidget(var numberedPoints) {
    return ListView.builder(
        shrinkWrap: true,
        itemCount: numberedPoints.length,
        itemBuilder: (BuildContext context, int index) {
          return numberedPoints[index].nodes.isEmpty
              ? Text('• ' + numberedPoints[index])
              : Row(
                  mainAxisAlignment: MainAxisAlignment.start,
                  mainAxisSize: MainAxisSize.max,
                  children: [
                    Text(index.toString() + '. '),
                    Expanded(
                      child: RichText(
                        text: TextSpan(
                          style: DefaultTextStyle.of(context).style,
                          children: _textSpanList(numberedPoints[index].nodes),
                        ),
                      ),
                    ),
                  ],
                );
        });
  }

  static _textSpanList(var pointsNodes) {
    List<TextSpan> list = [];
    for (int i = 0; i < pointsNodes.length; i++) {
      if (pointsNodes[i].toString().contains('<html b>')) {
        list.add(TextSpan(
            text: pointsNodes[i].text,
            style: TextStyle(fontWeight: FontWeight.bold)));
      } else {
        list.add(TextSpan(text: pointsNodes[i].text));
      }
    }
    return list;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:html_parser/html_widget.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'html_parser demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'html_parser demo home page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late String _htmlDoc;
  late List<Widget> _outputHtml;

  @override
  void initState() {
    super.initState();
    _htmlDoc = createHtmlDoc();
    _outputHtml = MyHtmlParser.parseHtmlToListOfTextWidgets(_htmlDoc);
  }

  String createHtmlDoc() {
    return '<h1>This package will help you display html document properly in your Flutter app.</h1>';
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              Text(
                'Html input: $_htmlDoc',
              ),
              Text(
                'Output: ',
              ),
              ListView.builder(
                shrinkWrap: true,
                itemCount: _outputHtml.length,
                itemBuilder: (contex, index) {
                  return _outputHtml.elementAt(index);
                },
              )
            ],
          ),
        ),
      ),
    );
  }
}


lib/html_widget.dart
library html_widget;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:html/parser.dart';

class MyHtmlParser {
  static List<Widget> parseHtmlToListOfTextWidgets(var text) {
    var data = parse(text.toString()).body;
    List<Widget> richText = [];
    for (int i = 0; i < data!.nodes.length; i++) {
      if (data.nodes[i].toString().contains('<html h1>')) {
        richText.add(_titleWidget(data.nodes[i].text));
      } else if (data.nodes[i].toString().contains('<html h2>')) {
        richText.add(_subtitleWidget(data.nodes[i].text));
      } else if (data.nodes[i].toString().contains('<html p>')) {
        richText.add(_paragraphWidget(data.nodes[i], data.nodes[i].text));
      } else if (data.nodes[i].toString().contains('<html ol>')) {
        richText.add(_numberedListWidget(data.nodes[i].nodes));
      } else if (data.nodes[i].toString().contains('<html ul>')) {
        richText.add(_bulletListWidget(data.nodes[i].nodes));
      } else if (data.nodes[i].toString().contains('<html b>')) {
        richText.add(_boldWidget(data.nodes[i].text));
      }
    }
    return richText;
  }

  static Widget _titleWidget(String? title) {
    return Text(title!,
        style: TextStyle(fontSize: 32, fontWeight: FontWeight.bold));
  }

  static Widget _subtitleWidget(String? subtitle) {
    return Text(subtitle!,
        style: TextStyle(fontSize: 24, fontWeight: FontWeight.bold));
  }

  static Widget _boldWidget(String? text) {
    return Text(text!, style: TextStyle(fontWeight: FontWeight.bold));
  }

  static Widget _paragraphWidget(var nodesList, String? text) {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 8.0),
      child: nodesList.nodes.isEmpty
          ? Text(text!)
          : RichText(
              text: TextSpan(
                style: TextStyle(
                  fontSize: 16.0,
                  color: Colors.black,
                ),
                children: _textSpanList(nodesList.nodes),
              ),
            ),
    );
  }

  static Widget _bulletListWidget(var bulletPoints) {
    return ListView.builder(
        shrinkWrap: true,
        itemCount: bulletPoints.length,
        itemBuilder: (BuildContext context, int index) {
          return bulletPoints[index].nodes.isEmpty
              ? Text('• ' + bulletPoints[index].text)
              : Row(
                  mainAxisAlignment: MainAxisAlignment.start,
                  mainAxisSize: MainAxisSize.max,
                  children: [
                    Text('• '),
                    Expanded(
                      child: RichText(
                        text: TextSpan(
                          style: DefaultTextStyle.of(context).style,
                          children: _textSpanList(bulletPoints[index].nodes),
                        ),
                      ),
                    ),
                  ],
                );
        });
  }

  static Widget _numberedListWidget(var numberedPoints) {
    return ListView.builder(
        shrinkWrap: true,
        itemCount: numberedPoints.length,
        itemBuilder: (BuildContext context, int index) {
          return numberedPoints[index].nodes.isEmpty
              ? Text('• ' + numberedPoints[index])
              : Row(
                  mainAxisAlignment: MainAxisAlignment.start,
                  mainAxisSize: MainAxisSize.max,
                  children: [
                    Text(index.toString() + '. '),
                    Expanded(
                      child: RichText(
                        text: TextSpan(
                          style: DefaultTextStyle.of(context).style,
                          children: _textSpanList(numberedPoints[index].nodes),
                        ),
                      ),
                    ),
                  ],
                );
        });
  }

  static _textSpanList(var pointsNodes) {
    List<TextSpan> list = [];
    for (int i = 0; i < pointsNodes.length; i++) {
      if (pointsNodes[i].toString().contains('<html b>')) {
        list.add(TextSpan(
            text: pointsNodes[i].text,
            style: TextStyle(fontWeight: FontWeight.bold)));
      } else {
        list.add(TextSpan(text: pointsNodes[i].text));
      }
    }
    return list;
  }
}
